AnnoDesigner/Helper/PolygonBoundaryFinderHelper.cs

```diff
@@ -0,0 +1,104 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Windows;
+
+namespace AnnoDesigner.Helper
+{
+    public static class PolygonBoundaryFinderHelper
+    {
+        private enum Direction
+        {
+            Up,
+            Left,
+            Down,
+            Right
+        }
+
+        private static Point GetLeftCell(Point point, Direction direction)
+        {
+            switch (direction)
+            {
+                case Direction.Up:
+                    return new Point(point.X - 1, point.Y - 1);
+                case Direction.Left:
+                    return new Point(point.X - 1, point.Y);
+                case Direction.Down:
+                    return point;
+                default:
+                    return new Point(point.X, point.Y - 1);
+            }
+        }
+
+        private static Point GetRightCell(Point point, Direction direction)
+        {
+            switch (direction)
+            {
+                case Direction.Up:
+                    return new Point(point.X, point.Y - 1);
+                case Direction.Left:
+                    return new Point(point.X - 1, point.Y - 1);
+                case Direction.Down:
+                    return new Point(point.X - 1, point.Y);
+                default:
+                    return point;
+            }
+        }
+
+        private static Point MoveForward(Point point, Direction direction)
+        {
+            switch (direction)
+            {
+                case Direction.Up:
+                    return new Point(point.X, point.Y - 1);
+                case Direction.Left:
+                    return new Point(point.X - 1, point.Y);
+                case Direction.Down:
+                    return new Point(point.X, point.Y + 1);
+                default:
+                    return new Point(point.X + 1, point.Y);
+            }
+        }
+
+        public static IList<Point> GetBoundaryPoints(ISet<Point> insidePoints)
+        {
+            var points = new List<Point>();
+
+            if (insidePoints.Count == 0)
+                return points;
+
+            var minPoint = insidePoints.Min(p => (p.X, p.Y));
+            var startPoint = new Point(minPoint.X, minPoint.Y);
+            var point = startPoint;
+            var direction = Direction.Down;
+            points.Add(point);
+
+            do
+            {
+                var left = GetLeftCell(point, direction);
+                var right = GetRightCell(point, direction);
+
+                if (insidePoints.Contains(left))
+                {
+                    if (insidePoints.Contains(right))// turn right
+                    {
+                        points.Add(point);
+
+                        direction = (Direction)(((int)direction + 3) % 4);
```

Comment:
I know this looks nicer, but as we are very hot on performance, would some kind of longer form switch statement be better for performance?

An addition and mod isn't exactly expensive, so it might not even be worth it, but just a bit of food for thought.

Reply:
You think one addition and one modulo operation on integer will be quicker that a switch? I'll test it but even if it turns out a switch is faster this seems like extreme case of premature optimalisation.